File: packages/typescript/src/catalogs/mexico/uma.ts
import { loadCatalogArray } from '../../utils/catalog-loader';
import { UMA } from '../../types';
import { SalariosMinimos } from './salarios-minimos';

/**
 * UMA (Unidad de Medida y Actualización) Catalog
 * Reference unit for fines, taxes, and economic calculations in Mexico
 * Replaced minimum wage as reference unit in 2017
 */
export class UMACatalog {
  private static _data: UMA[] | null = null;

  private static loadData(): void {
    if (this._data !== null) return;
    this._data = loadCatalogArray<UMA>('mexico/uma.json');
  }

  static getData(): UMA[] {
    this.loadData();
    return this._data!;
  }

  /**
   * Get UMA for a specific year
   */
  static getPorAño(año: number): UMA | undefined {
    const existente = this.getData().find((u) => u.año === año);
    if (existente) {
      return existente;
    }

    const equivalencia = SalariosMinimos.getUmaEquivalente(año, 'diario');
    if (equivalencia === undefined) {
      return undefined;
    }

    const salario = SalariosMinimos.getPorAño(año);
    const vigenciaInicio = salario?.vigencia_inicio ?? `${año}-01-01`;
    const vigenciaFin = salario?.vigencia_inicio ? `${año}-12-31` : `${año}-12-31`;
    const valorDiario = equivalencia;
    const valorMensual =
      SalariosMinimos.getUmaEquivalente(año, 'mensual') ?? Number((valorDiario * 30.4).toFixed(2));
    const valorAnual =
      SalariosMinimos.getUmaEquivalente(año, 'anual') ?? Number((valorDiario * 365).toFixed(2));

    return {
      año,
      vigencia_inicio: vigenciaInicio,
      vigencia_fin: vigenciaFin,
      valor_diario: Number(valorDiario.toFixed(2)),
      valor_mensual: Number(valorMensual.toFixed(2)),
      valor_anual: Number(valorAnual.toFixed(2)),
      moneda: salario?.moneda ?? 'MXN',
      publicacion_dof: salario?.vigencia_inicio ?? `${año}-01-01`,
      incremento_porcentual: null,
      notas: 'Equivalencia de UMA generada a partir del salario mínimo vigente antes de 2017',
    };
  }

  /**
   * Get current UMA (most recent year)
   */
  static getActual(): UMA {
    const sorted = [...this.getData()].sort((a, b) => b.año - a.año);
    return sorted[0];
  }

  /**
   * Get UMA value for a specific date
   */
  static getPorFecha(fecha: string | Date): UMA | undefined {
    const date = typeof fecha === 'string' ? new Date(fecha) : fecha;
    const sorted = [...this.getData()].sort(
      (a, b) => new Date(b.vigencia_inicio).getTime() - new Date(a.vigencia_inicio).getTime()
    );

    return (
      sorted.find((u) => {
        const inicio = new Date(u.vigencia_inicio);
        const fin = new Date(u.vigencia_fin);
        return date >= inicio && date <= fin;
      }) ?? this.getPorAño(date.getFullYear())
    );
  }

  /**
   * Get UMA value by type (diario, mensual, anual)
   */
  static getValor(
    año: number,
    tipo: 'diario' | 'mensual' | 'anual' = 'diario'
  ): number | undefined {
    const uma = this.getPorAño(año);
    if (!uma) return undefined;

    switch (tipo) {
      case 'diario':
        return uma.valor_diario;
      case 'mensual':
        return uma.valor_mensual;
      case 'anual':
        return uma.valor_anual;
    }
  }

  /**
   * Calculate amount in UMAs for a given monetary value
   */
  static calcularUMAs(
    monto: number,
    año: number,
    tipo: 'diario' | 'mensual' | 'anual' = 'diario'
  ): number | undefined {
    const valorUMA = this.getValor(año, tipo);
    return valorUMA ? monto / valorUMA : undefined;
  }

  /**
   * Calculate monetary value from UMAs
   */
  static calcularMonto(
    umas: number,
    año: number,
    tipo: 'diario' | 'mensual' | 'anual' = 'diario'
  ): number | undefined {
    const valorUMA = this.getValor(año, tipo);
    return valorUMA ? umas * valorUMA : undefined;
  }

  /**
   * Get historical evolution of UMA values
   */
  static getHistorico(añoInicio: number, añoFin: number): UMA[] {
    return this.getData()
      .filter((u) => u.año >= añoInicio && u.año <= añoFin)
      .sort((a, b) => a.año - b.año);
  }

  /**
   * Calculate percentage increase between two years
   */
  static calcularIncremento(añoInicial: number, añoFinal: number): number | undefined {
    const inicial = this.getValor(añoInicial);
    const final = this.getValor(añoFinal);

    if (!inicial || !final) return undefined;
    return ((final - inicial) / inicial) * 100;
  }

  /**
   * Get UMA increment percentage for a specific year
   */
  static getIncrementoAnual(año: number): number | null | undefined {
    const uma = this.getPorAño(año);
    return uma?.incremento_porcentual;
  }
}

File: packages/python/catalogmx/catalogs/mexico/uma.py
"""
UMA (Unidad de Medida y Actualización) Catalog

This module provides access to UMA values, which are used as reference
units for fines, taxes, and other obligations in Mexico.
"""

import json
from pathlib import Path

from .salarios_minimos import SalariosMinimos


class UMACatalog:
    """
    Catalog of UMA (Unidad de Medida y Actualización) values

    UMA is a reference economic unit used in Mexico for calculating
    fines, taxes, social security contributions, and other obligations.
    It was introduced in 2016 to replace the minimum wage as a reference unit.
    """

    _data: list[dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        """Load UMA data from JSON file"""
        if cls._data is None:
            # Path: catalogmx/packages/python/catalogmx/catalogs/mexico/uma.py
            # Target: catalogmx/packages/shared-data/mexico/uma.json
            current_file = Path(__file__)
            shared_data_path = (
                current_file.parent.parent.parent.parent.parent
                / "shared-data"
                / "mexico"
                / "uma.json"
            )

            with open(shared_data_path, encoding="utf-8") as f:
                cls._data = json.load(f)

    @classmethod
    def get_data(cls) -> list[dict]:
        """
        Get all UMA data

        :return: List of all UMA records
        """
        cls._load_data()
        return cls._data.copy()

    @classmethod
    def get_por_anio(cls, anio: int) -> dict | None:
        """
        Get UMA values for a specific year

        :param anio: Year (e.g., 2024)
        :return: UMA record or None if not found
        """
        cls._load_data()

        for record in cls._data:
            if record["año"] == anio:
                return record.copy()

        # Fallback to salary minimum equivalence for pre-2017 years
        salario = SalariosMinimos.get_por_anio(anio)
        if not salario:
            return None

        diario = salario.get("uma_equivalente_diario") or salario.get("resto_pais")
        if diario is None:
            return None

        mensual = salario.get("uma_equivalente_mensual")
        if mensual is None:
            mensual = round(diario * 30.4, 2)

        anual = salario.get("uma_equivalente_anual")
        if anual is None:
            anual = round(diario * 365, 2)

        return {
            "año": anio,
            "vigencia_inicio": salario.get("vigencia_inicio", f"{anio}-01-01"),
            "vigencia_fin": f"{anio}-12-31",
            "valor_diario": round(diario, 2),
            "valor_mensual": mensual,
            "valor_anual": anual,
            "moneda": salario.get("moneda", "MXN"),
            "publicacion_dof": salario.get("vigencia_inicio", f"{anio}-01-01"),
            "incremento_porcentual": None,
            "notas": "Equivalencia de UMA derivada del salario mínimo vigente antes de 2017",
        }

    @classmethod
    def get_actual(cls) -> dict | None:
        """
        Get current UMA values (latest year in data)

        :return: Latest UMA record
        """
        cls._load_data()

        if not cls._data:
            return None

        # Data is sorted by year descending, so first element is most recent
        return cls._data[0].copy()

    @classmethod
    def get_valor(cls, anio: int, tipo: str = "diario") -> float | None:
        """
        Get UMA value for specific year and type

        :param anio: Year
        :param tipo: Type of value ('diario', 'mensual', 'anual')
        :return: UMA value or None if not found
        """
        record = cls.get_por_anio(anio)
        if not record:
            return None

        tipo_map = {"diario": "valor_diario", "mensual": "valor_mensual", "anual": "valor_anual"}

        field = tipo_map.get(tipo.lower())
        if not field:
            return None

        return record.get(field)

    @classmethod
    def calcular_umas(cls, monto: float, anio: int, tipo: str = "diario") -> float | None:
        """
        Calculate how many UMAs a given amount represents

        :param monto: Amount in Mexican pesos
        :param anio: Year for UMA value
        :param tipo: Type of UMA ('diario', 'mensual', 'anual')
        :return: Number of UMAs or None if UMA value not found
        """
        valor_uma = cls.get_valor(anio, tipo)
        if not valor_uma:
            return None

        return monto / valor_uma

    @classmethod
    def calcular_monto(cls, umas: float, anio: int, tipo: str = "diario") -> float | None:
        """
        Calculate the peso amount for a given number of UMAs

        :param umas: Number of UMAs
        :param anio: Year for UMA value
        :param tipo: Type of UMA ('diario', 'mensual', 'anual')
        :return: Amount in Mexican pesos or None if UMA value not found
        """
        valor_uma = cls.get_valor(anio, tipo)
        if not valor_uma:
            return None

        return umas * valor_uma

    @classmethod
    def get_incremento(cls, anio: int) -> float | None:
        """
        Get percentage increment for a specific year

        :param anio: Year
        :return: Percentage increment or None if not found
        """
        record = cls.get_por_anio(anio)
        if not record:
            return None

        return record.get("incremento_porcentual")


# Convenience functions
def get_uma_actual() -> dict | None:
    """Get current UMA values"""
    return UMACatalog.get_actual()


def get_uma_por_anio(anio: int) -> dict | None:
    """Get UMA values for a specific year"""
    return UMACatalog.get_por_anio(anio)


def calcular_umas(monto: float, anio: int, tipo: str = "diario") -> float | None:
    """Calculate how many UMAs a given amount represents"""
    return UMACatalog.calcular_umas(monto, anio, tipo)


def calcular_monto(umas: float, anio: int, tipo: str = "diario") -> float | None:
    """Calculate peso amount for a given number of UMAs"""
    return UMACatalog.calcular_monto(umas, anio, tipo)


# Export commonly used functions and classes
__all__ = [
    "UMACatalog",
    "get_uma_actual",
    "get_uma_por_anio",
    "calcular_umas",
    "calcular_monto",
]
